Getting panic: inconsistent label cardinality when starting

I followed the supplied commands to `make` the project . While connecting to  my local Postgres Db,

I am getting : 

```
export DATA_SOURCE_NAME="postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable"

# ./postgres_exporter
INFO[0000] Established new database connection.          source=postgres_exporter.go:951
INFO[0000] Semantic Version Changed: 0.0.0 -> 9.4.1      source=postgres_exporter.go:904
panic: inconsistent label cardinality

goroutine 17 [running]:
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.MustNewConstMetric(0xc42021c770, 0x3, 0x0, 0xc42025b120, 0x2, 0x2, 0x0, 0x0)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/value.go:172 +0xb3
main.queryNamespaceMapping(0xc420184000, 0xc42014e280, 0x7c9e3f, 0x10, 0xc420129180, 0x2, 0x2, 0xc420139500, 0xc420139b90, 0x0, ...)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:853 +0x1200
main.queryNamespaceMappings(0xc420184000, 0xc42014e280, 0xc4201391a0, 0xc420139b90, 0x0)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:868 +0x321
main.(*Exporter).scrape(0xc420182000, 0xc420184000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:996 +0x58d
main.(*Exporter).Collect(0xc420182000, 0xc420184000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:747 +0x39
main.(*Exporter).Describe(0xc420182000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:740 +0xb0
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register.func1(0x97ab20, 0xc420182000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:218 +0x3b
created by postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:217 +0xbc
```
I am unsure about why the `panic: inconsistent label cardinality` error occurs.

Thank you for your reply. Here is the info.

```
$ ./postgres_exporter -log.level=debug
INFO[0000] Established new database connection.          source=postgres_exporter.go:951
DEBU[0000] Querying Postgres Version                     source=postgres_exporter.go:887
INFO[0000] Semantic Version Changed: 0.0.0 -> 9.4.1      source=postgres_exporter.go:904
DEBU[0000] procpid is being forced to discard due to version incompatibility.  source=postgres_exporter.go:496
DEBU[0000] Querying pg_setting view                      source=pg_setting.go:17
DEBU[0000] Querying namespace:  pg_stat_database_conflicts  source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_locks                 source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_replication      source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_activity         source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_bgwriter         source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_database         source=postgres_exporter.go:867
DEBU[0000] blks_icache_hit [1 template1]                 source=postgres_exporter.go:852
panic: inconsistent label cardinality

goroutine 17 [running]:
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.MustNewConstMetric(0xc42020ef50, 0x3, 0x0, 0xc42025b3a0, 0x2, 0x2, 0x0, 0x0)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/value.go:172 +0xb3
main.queryNamespaceMapping(0xc420186000, 0xc420150280, 0x7c9e3f, 0x10, 0xc420146fe0, 0x2, 0x2, 0xc420139470, 0xc420139bc0, 0x0, ...)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:853 +0x1200
main.queryNamespaceMappings(0xc420186000, 0xc420150280, 0xc4201391d0, 0xc420139bc0, 0x0)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:868 +0x321
main.(*Exporter).scrape(0xc420184000, 0xc420186000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:996 +0x58d
main.(*Exporter).Collect(0xc420184000, 0xc420186000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:747 +0x39
main.(*Exporter).Describe(0xc420184000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:740 +0xb0
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register.func1(0x97ab20, 0xc420184000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:218 +0x3b
created by postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:217 +0xbc
```

--- a/postgres_exporter.go
+++ b/postgres_exporter.go
@@ -125,6 +125,7 @@ type MetricMap struct {
 
 // TODO: revisit cu with the semver system
 func dumpMaps() {
+	// TODO: make this function part of the exporter
 	for name, cmap := range builtinMetricMaps {
 		query, ok := queryOverrides[name]
 		if !ok {
@@ -659,6 +660,10 @@ func dbToString(t interface{}) (string, bool) {
 
 // Exporter collects Postgres metrics. It implements prometheus.Collector.
 type Exporter struct {
+	// Holds a reference to the build in column mappings. Currently this is for testing purposes
+	// only, since it just points to the global.
+	builtinMetricMaps map[string]map[string]ColumnMapping
+
 	dsn              string
 	userQueriesPath  string
 	duration         prometheus.Gauge
@@ -685,8 +690,9 @@ type Exporter struct {
 // NewExporter returns a new PostgreSQL exporter for the provided DSN.
 func NewExporter(dsn string, userQueriesPath string) *Exporter {
 	return &Exporter{
-		dsn:             dsn,
-		userQueriesPath: userQueriesPath,
+		builtinMetricMaps: builtinMetricMaps,
+		dsn:               dsn,
+		userQueriesPath:   userQueriesPath,
 		duration: prometheus.NewGauge(prometheus.GaugeOpts{
 			Namespace: namespace,
 			Subsystem: exporter,
@@ -913,7 +919,7 @@ func (e *Exporter) checkMapVersions(ch chan<- prometheus.Metric, db *sql.DB) err
 		log.Infoln("Semantic Version Changed:", e.lastMapVersion.String(), "->", semanticVersion.String())
 		e.mappingMtx.Lock()
 
-		e.metricMap = makeDescMap(semanticVersion, builtinMetricMaps)
+		e.metricMap = makeDescMap(semanticVersion, e.builtinMetricMaps)
 		e.queryOverrides = makeQueryOverrideMap(semanticVersion, queryOverrides)
 		e.lastMapVersion = semanticVersion
 

--- a/postgres_exporter.go
+++ b/postgres_exporter.go
@@ -837,7 +837,8 @@ func queryNamespaceMapping(ch chan<- prometheus.Metric, db *sql.DB, namespace st
 				ch <- prometheus.MustNewConstMetric(metricMapping.desc, metricMapping.vtype, value, labels...)
 			} else {
 				// Unknown metric. Report as untyped if scan to float64 works, else note an error too.
-				desc := prometheus.NewDesc(fmt.Sprintf("%s_%s", namespace, columnName), fmt.Sprintf("Unknown metric from %s", namespace), nil, nil)
+				metricLabel := fmt.Sprintf("%s_%s", namespace, columnName)
+				desc := prometheus.NewDesc(metricLabel, fmt.Sprintf("Unknown metric from %s", namespace), mapping.labels, nil)
 
 				// Its not an error to fail here, since the values are
 				// unexpected anyway.

--- a/postgres_exporter.go
+++ b/postgres_exporter.go
@@ -19,6 +19,7 @@ import (
 	"gopkg.in/alecthomas/kingpin.v2"
 	"gopkg.in/yaml.v2"
 
+	"crypto/sha256"
 	"github.com/blang/semver"
 	_ "github.com/lib/pq"
 	"github.com/prometheus/client_golang/prometheus"
@@ -371,15 +372,10 @@ func makeQueryOverrideMap(pgVersion semver.Version, queryOverrides map[string][]
 // TODO: test code for all cu.
 // TODO: use proper struct type system
 // TODO: the YAML cu supports is "non-standard" - we should move away from it.
-func addQueries(queriesPath string, pgVersion semver.Version, exporterMap map[string]MetricMapNamespace, queryOverrideMap map[string]string) error {
+func addQueries(content []byte, pgVersion semver.Version, exporterMap map[string]MetricMapNamespace, queryOverrideMap map[string]string) error {
 	var extra map[string]interface{}
 
-	content, err := ioutil.ReadFile(queriesPath)
-	if err != nil {
-		return err
-	}
-
-	err = yaml.Unmarshal(content, &extra)
+	err := yaml.Unmarshal(content, &extra)
 	if err != nil {
 		return err
 	}
@@ -663,10 +659,11 @@ func dbToString(t interface{}) (string, bool) {
 
 // Exporter collects Postgres metrics. It implements prometheus.Collector.
 type Exporter struct {
-	dsn             string
-	userQueriesPath string
-	duration, error prometheus.Gauge
-	totalScrapes    prometheus.Counter
+	dsn              string
+	userQueriesPath  string
+	duration, error  prometheus.Gauge
+	userQueriesError *prometheus.GaugeVec
+	totalScrapes     prometheus.Counter
 
 	// dbDsn is the connection string used to establish the dbConnection
 	dbDsn string
@@ -706,6 +703,12 @@ func NewExporter(dsn string, userQueriesPath string) *Exporter {
 			Name:      "last_scrape_error",
 			Help:      "Whether the last scrape of metrics from PostgreSQL resulted in an error (1 for error, 0 for success).",
 		}),
+		userQueriesError: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: exporter,
+			Name:      "user_queries_load_error",
+			Help:      "Whether the user queries file was loaded and parsed successfully (1 for error, 0 for success).",
+		}, []string{"filename", "hashsum"}),
 		metricMap:      nil,
 		queryOverrides: nil,
 	}
@@ -746,6 +749,7 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
 	ch <- e.duration
 	ch <- e.totalScrapes
 	ch <- e.error
+	e.userQueriesError.Collect(ch)
 }
 
 func newDesc(subsystem, name, help string) *prometheus.Desc {
@@ -906,8 +910,24 @@ func (e *Exporter) checkMapVersions(ch chan<- prometheus.Metric, db *sql.DB) err
 		e.lastMapVersion = semanticVersion
 
 		if e.userQueriesPath != "" {
-			if err := addQueries(e.userQueriesPath, semanticVersion, e.metricMap, e.queryOverrides); err != nil {
+			// Clear the metric while a reload is happening
+			e.userQueriesError.Reset()
+
+			// Calculate the hashsum of the useQueries
+			userQueriesData, err := ioutil.ReadFile(e.userQueriesPath)
+			if err != nil {
 				log.Errorln("Failed to reload user queries:", e.userQueriesPath, err)
+				e.userQueriesError.WithLabelValues(e.userQueriesPath, "").Set(1)
+			} else {
+				hashsumStr := fmt.Sprintf("%x", sha256.Sum256(userQueriesData))
+
+				if err := addQueries(userQueriesData, semanticVersion, e.metricMap, e.queryOverrides); err != nil {
+					log.Errorln("Failed to reload user queries:", e.userQueriesPath, err)
+					e.userQueriesError.WithLabelValues(e.userQueriesPath, hashsumStr).Set(1)
+				} else {
+					// Mark user queries as successfully loaded
+					e.userQueriesError.WithLabelValues(e.userQueriesPath, hashsumStr).Set(0)
+				}
 			}
 		}
 

--- a/postgres_exporter.go
+++ b/postgres_exporter.go
@@ -661,7 +661,9 @@ func dbToString(t interface{}) (string, bool) {
 type Exporter struct {
 	dsn              string
 	userQueriesPath  string
-	duration, error  prometheus.Gauge
+	duration         prometheus.Gauge
+	error            prometheus.Gauge
+	connectionError  prometheus.Gauge
 	userQueriesError *prometheus.GaugeVec
 	totalScrapes     prometheus.Counter
 
@@ -702,6 +704,12 @@ func NewExporter(dsn string, userQueriesPath string) *Exporter {
 			Subsystem: exporter,
 			Name:      "last_scrape_error",
 			Help:      "Whether the last scrape of metrics from PostgreSQL resulted in an error (1 for error, 0 for success).",
+		}),
+		connectionError: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: exporter,
+			Name:      "last_scrape_connection_error",
+			Help:      "Whether the last scrape of metrics from PostgreSQL was able to connect to the server (1 for error, 0 for success).",
 		}),
 		userQueriesError: prometheus.NewGaugeVec(prometheus.GaugeOpts{
 			Namespace: namespace,
@@ -955,12 +963,16 @@ func (e *Exporter) getDB(conn string) (*sql.DB, error) {
 	if e.dbConnection == nil {
 		d, err := sql.Open("postgres", conn)
 		if err != nil {
+			e.connectionError.Set(1)
 			return nil, err
 		}
 		err = d.Ping()
 		if err != nil {
+			e.connectionError.Set(1)
 			return nil, err
 		}
+		e.connectionError.Set(0)
+
 		d.SetMaxOpenConns(1)
 		d.SetMaxIdleConns(1)
 		e.dbConnection = d

--- a/postgres_exporter_integration_test.go
+++ b/postgres_exporter_integration_test.go
@@ -95,3 +95,30 @@ func (s *IntegrationSuite) TestInvalidDsnDoesntCrash(c *C) {
 	c.Assert(exporter, NotNil)
 	exporter.scrape(ch)
 }
+
+// TestUnknownMetricParsingDoesntCrash deliberately deletes all the column maps out
+// of an exporter to test that the default metric handling code can cope with unknown columns.
+func (s *IntegrationSuite) TestUnknownMetricParsingDoesntCrash(c *C) {
+	// Setup a dummy channel to consume metrics
+	ch := make(chan prometheus.Metric, 100)
+	go func() {
+		for range ch {
+		}
+	}()
+
+	dsn := os.Getenv("DATA_SOURCE_NAME")
+	c.Assert(dsn, Not(Equals), "")
+
+	exporter := NewExporter(dsn, "")
+	c.Assert(exporter, NotNil)
+
+	// Convert the default maps into a list of empty maps.
+	emptyMaps := make(map[string]map[string]ColumnMapping, 0)
+	for k := range exporter.builtinMetricMaps {
+		emptyMaps[k] = map[string]ColumnMapping{}
+	}
+	exporter.builtinMetricMaps = emptyMaps
+
+	// scrape the exporter and make sure it works
+	exporter.scrape(ch)
+}

